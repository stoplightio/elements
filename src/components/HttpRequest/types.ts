export type ParamType = 'query' | 'path' | 'header';
