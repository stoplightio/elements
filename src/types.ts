import { JsonPath, NodeType } from '@stoplight/types';
import { IconName } from '@stoplight/ui-kit';
import { IContentsNode } from '@stoplight/ui-kit/TableOfContents/types';
import { Edge, Node } from 'vis';
export { IComponentMapping } from '@stoplight/markdown-viewer';

export interface IDeserializedSrn {
  service: string;
  org: string;
  project: string;
  uri: string;
}

export interface IChange {
  createdAt: string;
  semver: string;
  message: string;
}

export interface IDiagnostic {
  location: { uri: string; jsonPath?: JsonPath };
  message: string;
}

export interface INodeInfo<D = unknown> {
  id: number;
  type: NodeType;
  name: string;
  srn: string;
  data: D;

  errors?: IDiagnostic[];
  changes?: IChange[];
  version?: string;
  versions?: string[];
  tags?: string[];
}

export interface IProjectNode {
  type: NodeType;
  srn: string;
  name: string;
  id: number | string;

  latestVersion?: string;
  versions?: string[];
  tags?: string[];
  data?: string;
  summary?: string;
}

export interface IContentsNodeWithId extends IContentsNode {
  id: number | string;
}

export type ProjectNodeWithUri = IProjectNode & { uri: string };

export type IconMapType = NodeType | 'group' | 'divider' | 'item';
export type NodeIconMapping = { [type in IconMapType]?: IconName };

export type DocsNodeType = NodeType | 'json_schema' | 'http';

export interface IPageHeading {
  id: string;
  title: string;
  depth: number;
}

export interface IPaginatedResponse<T> {
  items: T[];
  pageInfo: {
    hasNextPage: boolean;
    hasPreviousPage: boolean;
    startCursor: string;
    endCursor: string;
  };
  totalCount: number;
}

export interface IVisGraph {
  nodes: Node[];
  edges: Edge[];
}

export interface INodeGraph {
  nodes: IGraphNode[];
  edges: IGraphEdge[];
}

export interface IGraphNode {
  groupNodeId: number;
  name: string;
  srn: string;
  uri: string;
  depth: number;
  type: NodeType | string;
  version: string;

  projectName: string;
  groupSlug: string;
}

export interface IGraphEdge {
  fromGroupNodeId: number;
  fromPath: string;
  toGroupNodeId: number;
  toPath: string;
}
