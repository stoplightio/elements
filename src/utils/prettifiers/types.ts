export type Prettifier<T> = (content: T) => string | Promise<string>;
