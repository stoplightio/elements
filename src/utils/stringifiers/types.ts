export type Stringifier<T> = (content: T) => string;
