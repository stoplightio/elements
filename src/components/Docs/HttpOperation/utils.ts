import { INodeExample, INodeExternalExample } from '@stoplight/types';
import { isObject } from 'lodash';

export function getExamplesObject(examples: Array<INodeExample | INodeExternalExample>) {
  return examples.reduce((collection, item) => {
    const value = 'externalValue' in item ? item.externalValue : item.value;
    if (value) {
      collection[item.key] = value;
    }

    return collection;
  }, {});
}

export function getExamplesFromSchema(data: unknown) {
  // `examples` are available in JSON Schema v6 and v7. For v4 we can use `x-examples`.
  // `example` is not supported by any version but we can use `x-example` extension.
  return isObject(data)
    ? {
        ...(isObject(data['x-examples']) && { ...data['x-examples'] }),
        ...(isObject(data['examples']) && { ...data['examples'] }),
        ...('x-example' in data && { default: data['x-example'] }),
      }
    : void 0;
}
