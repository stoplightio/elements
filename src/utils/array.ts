export function flatMap<T, U>(array: T[], callbackfn: (value: T, index: number, array: T[]) => U[]): U[] {
  return Array.prototype.concat(...array.map(callbackfn));
}
