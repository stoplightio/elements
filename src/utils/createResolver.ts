import { Resolver } from '@stoplight/json-ref-resolver';
import { deserializeSrn, dirname, resolve, serializeSrn } from '@stoplight/path';
import { parse } from '@stoplight/yaml';
import axios from 'axios';
import * as URI from 'urijs';

export function createResolver(host: string, srn?: string) {
  return new Resolver({
    dereferenceInline: true,
    dereferenceRemote: true,

    resolvers: {
      https: httpResolver,
      http: httpResolver,
      file: remoteFileResolver(host, srn),
    },

    async parseResolveResult(opts) {
      if (typeof opts.result === 'string') {
        try {
          opts.result = parse(opts.result);
        } catch (e) {
          // noop, probably not json/yaml
        }
      }

      return opts;
    },
  });
}

/**
 * Resolves a $ref to another file in the same project to it's export URL in the Stoplight API
 *
 * @param host Stoplight API host
 * @param srn The active node's SRN
 */
function remoteFileResolver(host: string, srn?: string) {
  const { shortcode, orgSlug, projectSlug, uri } = deserializeSrn(srn || '');

  return {
    resolve(ref: uri.URI) {
      // If we don't have a URI, there's no way we can resolve the $ref's uri
      if (!uri) {
        throw new Error(`Failed to resolve ${ref}`);
      }

      // Rebuild the SRN with the $ref's uri
      const refSrn = serializeSrn({
        shortcode,
        orgSlug,
        projectSlug,
        uri: resolve(dirname(uri), String(ref)),
      });

      // Use the http resolver to resolve the node's raw export
      return httpResolver.resolve(new URI(`${host}/nodes.raw?srn=${refSrn}`));
    },
  };
}

const httpResolver = {
  async resolve(ref: uri.URI) {
    const res = await axios.get(String(ref));
    return res.data;
  },
};
