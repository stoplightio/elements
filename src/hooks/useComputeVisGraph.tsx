import { pointerToPath } from '@stoplight/json';
import { IGraphNodeData, IResolveResult } from '@stoplight/json-ref-resolver/types';
import { last } from 'lodash';
import * as React from 'react';
import { IVisGraph, IVisGraphEdge } from '../types';
import { getNodeTitle } from '../utils/node';

export function useComputeVisGraph(graph?: IResolveResult['graph'], rootName?: string, activeNodeId?: string) {
  return React.useMemo(() => computeVisGraph(graph, rootName, activeNodeId), [graph, rootName, activeNodeId]);
}

export function computeVisGraph(graph?: IResolveResult['graph'], rootName?: string, activeNodeId?: string) {
  const visGraph: IVisGraph = {
    nodes: [],
    edges: [],
  };

  if (!graph) return visGraph;

  // Loop over graph nodes and add them to the visgraph
  for (const id of graph.overallOrder()) {
    const isRootNode = id === 'root';
    const encodedId = encodeURI(id);
    const node = graph.getNodeData(id);

    let color = '#f5f8fa';
    if (isRootNode) {
      color = '#ef932b';
    } else if (activeNodeId === encodedId) {
      color = '#66b1e7';
    }

    visGraph.nodes.push({
      id: encodedId,
      label: isRootNode && rootName ? rootName : getNodeTitle(encodedId, node.data),
      color,
      font: {
        color: isRootNode || activeNodeId === encodedId ? '#ffffff' : '#10161a',
      },
    });

    // Add node edges
    visGraph.edges = visGraph.edges.concat(getEdgesFromRefMap(encodedId, node.refMap));
  }

  return visGraph;
}

function getEdgesFromRefMap(nodeId: string, refMap: IGraphNodeData['refMap']) {
  const edges: IVisGraphEdge[] = [];
  const edgeMap = {};

  // Loop over the ref map to combine duplicate edges
  for (const propertyPath in refMap) {
    if (!refMap.hasOwnProperty(propertyPath)) continue;

    const propertyKey = String(last(pointerToPath(propertyPath)));
    // numbered properties are probably part of a combiner so just show "reference" instead
    const propertyName = isNaN(parseInt(propertyKey)) ? propertyKey : 'reference';

    const encodedRef = encodeURI(refMap[propertyPath]);

    if (edgeMap.hasOwnProperty(encodedRef)) {
      // Don't add this property key if it already exists
      if (!edgeMap[encodedRef].includes(propertyName)) {
        edgeMap[encodedRef].push(propertyName);
      }
    } else {
      edgeMap[encodedRef] = [propertyName];
    }
  }

  // Loop over edges and add them to the graph
  for (const targetId in edgeMap) {
    if (!edgeMap.hasOwnProperty(targetId)) continue;

    edges.push({
      from: nodeId,
      to: targetId,
      label: edgeMap[targetId].length ? edgeMap[targetId].join(',\n') : 'reference',
      font: {
        align: edgeMap[targetId].length > 1 ? 'horizontal' : 'top',
      },
    });
  }

  return edges;
}
