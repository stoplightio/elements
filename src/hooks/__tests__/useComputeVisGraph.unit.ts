const nock = require('nock');
import axios from 'axios';
const simpleSchema = require('../../__fixtures__/schemas/simple.json');
const todoFullSchema = require('../../__fixtures__/schemas/todo-full.v1.json');
const todoPartialSchema = require('../../__fixtures__/schemas/todo-partial.v1.json');
const userSchema = require('../../__fixtures__/schemas/user.v1.json');

import { IResolveResult } from '@stoplight/json-ref-resolver/types';
import { INodeInfo } from '../../types';
import { createResolver } from '../../utils/createResolver';
import { computeVisGraph } from '../useComputeVisGraph';

beforeEach(() => {
  nock('https://stoplight.io/api')
    .persist()
    .get('/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-full.v1.json')
    .reply(200, todoFullSchema)
    .get('/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-partial.v1.json')
    .reply(200, todoPartialSchema)
    .get('/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/user.v1.json')
    .reply(200, userSchema);
});

afterEach(() => {
  nock.cleanAll();
});

describe('computeVisGraph', () => {
  test('it works with a simple example', async () => {
    const client = axios.create({
      baseURL: 'https://stoplight.io/api',
    });
    const { graph } = await createResolver(client).resolve(simpleSchema);

    expect(computeVisGraph({ srn: '' } as INodeInfo, graph, '')).toMatchSnapshot();
  });

  test('it handles circular references to the root node', async () => {
    const nodeData = {
      root: {
        data: {},
        refMap: {
          '#/property/path': 'http://stoplight.io/nodes.raw?srn=some/other/node',
        },
      },
      'http://stoplight.io/nodes.raw?srn=some/other/node': {
        data: {},
        refMap: {
          '#/property/path': 'http://stoplight.io/nodes.raw?srn=root/node/srn',
        },
      },
      'http://stoplight.io/nodes.raw?srn=root/node/srn': {
        data: {},
        refMap: {
          '#/property/path': 'http://stoplight.io/nodes.raw?srn=some/other/node',
        },
      },
    };

    const graph = {
      getNodeData: (id: string) => nodeData[id],
    } as IResolveResult['graph'];

    expect(computeVisGraph({ srn: 'root/node/srn' } as INodeInfo, graph, 'root')).toEqual({
      nodes: [
        {
          id: 'root',
          level: 0,
          color: '#ef932b',
          font: {
            color: '#ffffff',
          },
        },
        {
          id: 'http://stoplight.io/nodes.raw?srn=some/other/node',
          level: 1,
          label: 'Node',
          color: '#f5f8fa',
          font: {
            color: '#10161a',
          },
        },
      ],
      edges: [
        {
          from: 'http://stoplight.io/nodes.raw?srn=some/other/node',
          to: 'root',
          label: '',
          title: 'path',
          color: {
            color: '#cfd9e0',
            opacity: 0.8,
          },
          font: {
            align: 'top',
          },
        },
        {
          from: 'root',
          to: 'http://stoplight.io/nodes.raw?srn=some/other/node',
          label: '',
          title: 'path',
          color: {
            color: '#cfd9e0',
            opacity: 0.8,
          },
          font: {
            align: 'top',
          },
        },
      ],
    });
  });
});
