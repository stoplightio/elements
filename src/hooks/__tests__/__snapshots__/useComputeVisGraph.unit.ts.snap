// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeVisGraph it works with a simple example 1`] = `
Object {
  "edges": Array [
    Object {
      "font": Object {
        "align": "top",
      },
      "from": "https://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-full.v1.json",
      "label": "reference",
      "to": "http://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-partial.v1.json",
    },
    Object {
      "font": Object {
        "align": "top",
      },
      "from": "https://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-full.v1.json",
      "label": "user",
      "to": "http://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/user.v1.json",
    },
    Object {
      "font": Object {
        "align": "top",
      },
      "from": "root",
      "label": "todo",
      "to": "https://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-full.v1.json",
    },
    Object {
      "font": Object {
        "align": "top",
      },
      "from": "root",
      "label": "name",
      "to": "#/definitions/name",
    },
  ],
  "nodes": Array [
    Object {
      "color": "#f5f8fa",
      "font": Object {
        "color": "#10161a",
      },
      "id": "#/definitions/name",
      "label": "Name",
    },
    Object {
      "color": "#f5f8fa",
      "font": Object {
        "color": "#10161a",
      },
      "id": "http://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-partial.v1.json",
      "label": "Todo Partial",
    },
    Object {
      "color": "#f5f8fa",
      "font": Object {
        "color": "#10161a",
      },
      "id": "http://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/user.v1.json",
      "label": "User",
    },
    Object {
      "color": "#f5f8fa",
      "font": Object {
        "color": "#10161a",
      },
      "id": "https://stoplight.io/api/nodes.raw?srn=gh/stoplightio/studio-demo/reference/todos/models/todo-full.v1.json",
      "label": "Todo Full",
    },
    Object {
      "color": "#ef932b",
      "font": Object {
        "color": "#ffffff",
      },
      "id": "root",
      "label": "Root",
    },
  ],
}
`;
